src/core/tsi/alts/handshaker/alts_handshaker_client.cc

```diff
@@ -358,6 +335,164 @@ static tsi_result make_grpc_call(alts_handshaker_client* c, bool is_start) {
   return TSI_OK;
 }
 
+// TODO(apolcyn): remove this global queue when we can safely rely
+// on a MAX_CONCURRENT_STREAMS setting in the ALTS handshake server to
+// limit the number of concurrent handshakes.
+namespace {
+
+class HandshakeQueue {
+ public:
+  HandshakeQueue(size_t max_outstanding_handshakes)
+      : max_outstanding_handshakes_(max_outstanding_handshakes) {}
+
+  ~HandshakeQueue() { GRPC_COMBINER_UNREF(combiner_, "~HandshakeQueue"); }
+
+  void EnqueueHandshakeAndMaybeStartSome(alts_grpc_handshaker_client* client) {
+    Arg* arg = grpc_core::New<Arg>();
+    arg->client = client;
+    arg->self = this;
+    GRPC_CLOSURE_INIT(&arg->closure, EnqueueHandshakeAndMaybeStartSomeLocked,
+                      arg, nullptr);
+    combiner_->Run(&arg->closure, GRPC_ERROR_NONE);
```

Comment:
Do we switch to `grpc_core::Combiner` from `grpc_schedule_on_exec_ctx` in ALTS code?

Reply:
Because the only things we need to synchronize here are the fields of `HandshakeQueue` (which are only accessed within `EnqueueHandshakeAndMaybeStartSomeLocked`), this is the only thing that needs to use a combiner. Switching to use a combiner throughout the rest of the handshake isn't needed for this change.

Reply:
Acknowledged.